tsl/src/continuous_aggs/invalidation_threshold.c

```diff
@@ -66,47 +67,67 @@
 
 typedef struct InvalidationThresholdData
 {
-	int64 threshold;
-	bool was_updated;
+	const ContinuousAgg *cagg;
+	const InternalTimeRange *refresh_window;
+	int64 computed_invalidation_threshold;
 } InvalidationThresholdData;
 
-static ScanTupleResult scan_update_invalidation_threshold(TupleInfo *ti, void *data);
-static ScanTupleResult invalidation_threshold_tuple_found(TupleInfo *ti, void *data);
-
 static ScanTupleResult
-scan_update_invalidation_threshold(TupleInfo *ti, void *data)
+invalidation_threshold_scan_update(TupleInfo *ti, void *const data)
 {
-	InvalidationThresholdData *invthresh = data;
-	bool should_free;
-	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
-	Form_continuous_aggs_invalidation_threshold form =
-		(Form_continuous_aggs_invalidation_threshold) GETSTRUCT(tuple);
+	InvalidationThresholdData *invthresh = (InvalidationThresholdData *) data;
+
+	if (ti->lockresult != TM_Ok)
+		return SCAN_CONTINUE;
+
+	bool isnull;
+	Datum datum =
+		slot_getattr(ti->slot, Anum_continuous_aggs_invalidation_threshold_watermark, &isnull);
+	int64 current_invalidation_threshold = DatumGetInt64(datum);
+
+	/* Compute new invalidation threshold. Note that this computation caps the
+	 * threshold at the end of the last bucket that holds data in the
+	 * underlying hypertable. */
+	invthresh->computed_invalidation_threshold =
+		invalidation_threshold_compute(invthresh->cagg, invthresh->refresh_window);
 
-	if (invthresh->threshold > form->watermark)
+	if (invthresh->computed_invalidation_threshold > current_invalidation_threshold)
 	{
-		HeapTuple new_tuple = heap_copytuple(tuple);
-		form = (Form_continuous_aggs_invalidation_threshold) GETSTRUCT(new_tuple);
+		bool nulls[Natts_continuous_agg];
+		Datum values[Natts_continuous_agg];
+		bool doReplace[Natts_continuous_agg] = { false };
```

Comment:
Nit (not sure we are still following the old variable name standards, but I will assume we are):
```suggestion
		bool do_replace[Natts_continuous_agg] = { false };
```